cam/encoder.go

```diff
@@ -27,91 +27,140 @@ type encoder struct {
-	enc.codecCtx.time_base = C.AVRational{num: 1, den: C.int(framerate)}
-	enc.codecCtx.width = C.int(width)
-	enc.codecCtx.height = C.int(height)
-
+	e.codecCtx.bit_rate = C.int64_t(e.bitrate)
+	e.codecCtx.pix_fmt = C.AV_PIX_FMT_YUV422P
+	e.codecCtx.time_base = C.AVRational{num: 1, den: C.int(e.framerate)}
+	e.codecCtx.width = C.int(width)
+	e.codecCtx.height = C.int(height)
 	// TODO(seanp): Do we want b frames? This could make it more complicated to split clips.
-	enc.codecCtx.max_b_frames = 0
-	presetCStr := C.CString(preset)
+	e.codecCtx.max_b_frames = 0
+	presetCStr := C.CString(e.preset)
 	tuneCStr := C.CString("zerolatency")
 	defer C.free(unsafe.Pointer(presetCStr))
 	defer C.free(unsafe.Pointer(tuneCStr))
-
 	// The user can set the preset and tune for the encoder. This affects the
 	// encoding speed and quality. See https://trac.ffmpeg.org/wiki/Encode/H.264
 	// for more information.
 	var opts *C.AVDictionary
 	defer C.av_dict_free(&opts)
 	ret := C.av_dict_set(&opts, C.CString("preset"), presetCStr, 0)
 	if ret < 0 {
-		return nil, fmt.Errorf("av_dict_set failed: %s", ffmpegError(ret))
+		return fmt.Errorf("av_dict_set failed: %s", ffmpegError(ret))
 	}
 	ret = C.av_dict_set(&opts, C.CString("tune"), tuneCStr, 0)
 	if ret < 0 {
-		return nil, fmt.Errorf("av_dict_set failed: %s", ffmpegError(ret))
+		return fmt.Errorf("av_dict_set failed: %s", ffmpegError(ret))
 	}
-
-	ret = C.avcodec_open2(enc.codecCtx, codec, &opts)
+	ret = C.avcodec_open2(e.codecCtx, codec, &opts)
 	if ret < 0 {
-		return nil, fmt.Errorf("avcodec_open2: %s", ffmpegError(ret))
+		return fmt.Errorf("avcodec_open2: %s", ffmpegError(ret))
 	}
-
 	srcFrame := C.av_frame_alloc()
 	if srcFrame == nil {
-		C.avcodec_close(enc.codecCtx)
-		return nil, errors.New("could not allocate source frame")
+		C.avcodec_close(e.codecCtx)
+		return errors.New("could not allocate source frame")
 	}
-	srcFrame.width = enc.codecCtx.width
-	srcFrame.height = enc.codecCtx.height
-	srcFrame.format = C.int(enc.codecCtx.pix_fmt)
-	enc.srcFrame = srcFrame
-
-	return enc, nil
+	srcFrame.width = e.codecCtx.width
+	srcFrame.height = e.codecCtx.height
+	srcFrame.format = C.int(e.codecCtx.pix_fmt)
+	e.srcFrame = srcFrame
+	return nil
 }
 
 // encode encodes the given frame and returns the encoded data
 // in bytes along with the PTS and DTS timestamps.
 // PTS is calculated based on the frame count and source framerate.
 // If the polling loop is not running at the source framerate, the
 // PTS will lag behind actual run time.
-func (e *encoder) encode(frame image.Image) ([]byte, int64, int64, error) {
+func (e *encoder) encode(frame image.Image) (encodeResult, error) {
+	var result encodeResult
+	dy, dx := frame.Bounds().Dy(), frame.Bounds().Dx()
+	if e.codecCtx == nil || dy != int(e.codecCtx.height) || dx != int(e.codecCtx.width) {
+		e.logger.Infof("Initializing encoder with frame dimensions %dx%d", dx, dy)
+		err := e.initialize(dx, dy)
+		if err != nil {
+			return result, err
```

Comment:
is this meant to be an empty result or a nil? might be more readable to make result an explicit `result := encoderResult{}` rather than a `var result encoderResult` because I do not know if that initializes to nil from reading this code.

Reply:
Good call to be explicit here with default values for encodeResult so it is easier to read in early return cases. Changed.